vim-engine/src/main/kotlin/com/maddyhome/idea/vim/api/VimMarkServiceBase.kt

```diff
@@ -0,0 +1,605 @@
+    val mark = createMark(caret, markChar, offset) ?: return false
+    return setMark(caret, mark)
+  }
+
+  protected open fun createGlobalMark(editor: VimEditor, char: Char, offset: Int): Mark? {
+    val markChar = if (char == '`') '\'' else char
+    if (!markChar.isGlobalMark()) return null
+    if (!markChar.isValid(VimMarkService.Operation.SET, editor.primaryCaret())) return null
+    val position = editor.offsetToBufferPosition(offset)
+    val path = editor.getPath() ?: return null
+    return VimMark(markChar, position.line, position.column, path, editor.extractProtocol())
+  }
+
+  override fun setGlobalMark(editor: VimEditor, char: Char, offset: Int): Boolean {
+    val mark = createGlobalMark(editor, char, offset) ?: return false
+    return setGlobalMark(editor, mark)
+  }
+
+  override fun setGlobalMark(editor: VimEditor, mark: Mark): Boolean {
+    if (!mark.key.isValid(VimMarkService.Operation.SET, editor.primaryCaret())) return false
+    globalMarks[mark.key] = mark
+    return true
+  }
+
+  override fun setMarkForCaret(caret: ImmutableVimCaret, char: Char, offset: Int): Boolean {
+    val markChar = if (char == '`') '\'' else char
+    if (!markChar.isValid(VimMarkService.Operation.SET, caret)) return false
+    return when {
+      markChar.isGlobalMark() -> if (caret.isPrimary) setGlobalMark(caret.editor, char, offset) else false
+      markChar.isLocalMark() -> setMark(caret, char, offset)
+      else -> false
+    }
+  }
+
+  override fun setVisualSelectionMarks(editor: VimEditor) {
+    if (!editor.inVisualMode) return
+    val selectionType = SelectionType.fromSubMode(editor.subMode)
+    editor.carets()
+      .forEach {
+        val start = editor.offsetToBufferPosition(it.vimSelectionStart)
+        val end = editor.offsetToBufferPosition(it.offset.point)
+        it.lastSelectionInfo = SelectionInfo(start, end, selectionType)
+      }
+  }
+
+  override fun setVisualSelectionMarks(caret: ImmutableVimCaret, range: TextRange) {
+    setMark(caret, SELECTION_START_MARK, range.startOffset)
+    setMark(caret, SELECTION_END_MARK, range.endOffset)
+  }
+
+  override fun getVisualSelectionMarks(caret: ImmutableVimCaret): TextRange? {
+    return getMarksRange(caret, SELECTION_START_MARK, SELECTION_END_MARK)
+  }
+
+  override fun setChangeMarks(caret: ImmutableVimCaret, range: TextRange) {
+    setMark(caret, CHANGE_START_MARK, range.startOffset)
+    setMark(caret, CHANGE_END_MARK, range.endOffset - 1)
+  }
+
+  override fun getChangeMarks(caret: ImmutableVimCaret): TextRange? {
+    return getMarksRange(caret, CHANGE_START_MARK, CHANGE_END_MARK)
+  }
+
+  private fun getMarksRange(caret: ImmutableVimCaret, startMark: Char, endMark: Char): TextRange? {
+    val editor = caret.editor
+    val startOffset = getLocalMark(caret, startMark)
+      ?.offset(editor) ?: return null
+    val endOffset = getLocalMark(caret, endMark)
+      ?.offset(editor) ?: return null
+    return TextRange(startOffset, endOffset + 1)
```

Comment:
Are you sure about this `+1` magic constant? Do we always keep the end offset as exclusive? Do we always need it inclusive? Does `endOffset + 1` always < file size? Is it possible that `startOffset > endOffset`?